pytest_rerunfailures.py

```diff
@@ -250,12 +255,155 @@ def _should_not_rerun(item, report, reruns):
+        node.workerinput['sock_port'] = node.config.failures_db.sock_port
+
+
+    def pytest_handlecrashitem(crashitem, report, sched):
+        """
+        Return the crashitem from pending and collection.
+        """
+        db = sched.config.failures_db
+        reruns = db.get_test_reruns(crashitem)
+        if db.get_test_failures(crashitem) < reruns:
+            sched.mark_test_pending(crashitem)
+            report.outcome = "rerun"
+
+        db.add_test_failure(crashitem)
+
+
+class StatusDB(object):
+    def __init__(self):
+        self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        self.sock.setblocking(1)
+        self.delim = '\n'.encode()
+        self.hmap = {}
+
+    def _hash(self, crashitem: str) -> str:
+        if crashitem not in self.hmap:
+            self.hmap[crashitem] = hashlib.sha1(
+                crashitem.encode(),
+            ).hexdigest()[:10]
+
+        return self.hmap[crashitem]        
+
+    def add_test_failure(self, crashitem):
+        hash = self._hash(crashitem)
+        failures = self._get(hash, 'f')
+        failures += 1
+        self._set(hash, 'f', failures)
+
+    def get_test_failures(self, crashitem):
+        hash = self._hash(crashitem)
+        return self._get(hash, 'f')
+
+    def set_test_reruns(self, crashitem, reruns):
+        hash = self._hash(crashitem)
+        self._set(hash, 'r', reruns)
+
+    def get_test_reruns(self, crashitem):
+        hash = self._hash(crashitem)
+        return self._get(hash, 'r')
+
+    def _set(self, i:str, k:str, v:int):
+        pass
+    
+    def _get(self, i:str, k:str) -> int:
+        return 0
+
+    def _sock_recv(self, conn) -> str:
+        buf = b''
+        while True:
+            b = conn.recv(1)
+            if b == self.delim:
+                break
+            buf += b
+
+        return buf.decode()
+
+    def _sock_send(self, conn, msg: str):
+        conn.send(msg.encode() + self.delim)
+
+
+class ServerStatusDB(StatusDB):
+    def __init__(self):
+        super().__init__()
+        self.sock.bind(('', 0))
+        self.sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+
+        self.rerunfailures_db = {}
+        t = threading.Thread(target=self.run_server, daemon=True)
+        t.start()
+
+    @property
+    def sock_port(self):
+        return self.sock.getsockname()[1]
+
+    def run_server(self):
+        self.sock.listen()
+        while True:
+            conn, _ = self.sock.accept()
+            t = threading.Thread(target=self.run_connection, args=(conn,), daemon=True)
+            t.start()
+
+    def run_connection(self, conn):
+        with suppress(ConnectionError):
+            while True:
+                op, i, k, v = self._sock_recv(conn).split('|')
+                if op == 'set':
+                    self._set(i, k, int(v))
+                elif op == 'get':
+                    self._sock_send(conn, str(self._get(i, k)))
+
+    def _set(self, i:str, k:str, v:int):
```

Comment:
I do not understand the meaning of the parameters, maybe longer variable names could clarify.

Reply:
`i` is a hash of the test name (`test_file.py::test_my_test`), `k` is `f` for failures or `r` for reruns. v is the number of failures or reruns (an int). before we run anything in the test (including setup), the worker needs to record in these in the master so that the master can decide whether the test needs to be rerun if it crashes. the worker must communicate this information because the number of reruns can only be computed by collecting tests, which can only be done in the worker. Again, if xdist is not found this this functionality does nothing.

Reply:
To clarify, the StatusDB is an in-memory database stored on the master that records, for each test, the number of failures and the number of reruns. The Status DB is accessed on the worker through a socket connection that binds to a random free port (assigned by the OS). This will allow (eventually) the remote xdist feature to work (when someone requires it).

Reply:
> `i` is a hash of the test name (`test_file.py::test_my_test`), `k` is `f` for failures or `r` for reruns. v is the number of failures or reruns (an int). before we run anything in the test (including setup), the worker needs to record in these in the master so that the master can decide whether the test needs to be rerun if it crashes. the worker must communicate this information because the number of reruns can only be computed by collecting tests, which can only be done in the worker. Again, if xdist is not found this this functionality does nothing.

It would be important to have this explanation in the code.